pkg/frontend/querymiddleware/error_caching_test.go

```diff
@@ -0,0 +1,221 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/user"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/util/test"
+)
+
+func TestErrorCachingHandler_Do(t *testing.T) {
+	newDefaultRequest := func() *PrometheusRangeQueryRequest {
+		return &PrometheusRangeQueryRequest{
+			queryExpr: parseQuery(t, "up"),
+			start:     100,
+			end:       200,
+			step:      10,
+		}
+	}
+
+	runHandler := func(ctx context.Context, inner MetricsQueryHandler, c cache.Cache, req MetricsQueryRequest) (Response, error) {
+		limits := &mockLimits{resultsCacheTTLForErrors: time.Minute}
+		middleware := newErrorCachingMiddleware(c, limits, resultsCacheEnabledByOption, test.NewTestingLogger(t), prometheus.NewPedanticRegistry())
+		handler := middleware.Wrap(inner)
+		return handler.Do(ctx, req)
+	}
+
+	t.Run("no user set", func(t *testing.T) {
+		c := cache.NewInstrumentedMockCache()
+
+		innerRes := newEmptyPrometheusResponse()
+		inner := &mockHandler{}
+		inner.On("Do", mock.Anything, mock.Anything).Return(innerRes, nil)
+
+		ctx := context.Background()
+		req := newDefaultRequest()
+		res, err := runHandler(ctx, inner, c, req)
+
+		require.NoError(t, err)
+		require.Equal(t, innerRes, res)
+		require.Equal(t, 0, c.CountFetchCalls())
+		require.Equal(t, 0, c.CountStoreCalls())
+	})
+
+	t.Run("disabled by option", func(t *testing.T) {
+		c := cache.NewInstrumentedMockCache()
+
+		innerRes := newEmptyPrometheusResponse()
+		inner := &mockHandler{}
+		inner.On("Do", mock.Anything, mock.Anything).Return(innerRes, nil)
+
+		ctx := user.InjectOrgID(context.Background(), "1234")
+		req := newDefaultRequest()
+		req.options = Options{
+			CacheDisabled: true,
+		}
+		res, err := runHandler(ctx, inner, c, req)
+
+		require.NoError(t, err)
+		require.Equal(t, innerRes, res)
+		require.Equal(t, 0, c.CountFetchCalls())
+		require.Equal(t, 0, c.CountStoreCalls())
+	})
+
+	t.Run("cache hit", func(t *testing.T) {
+		c := cache.NewInstrumentedMockCache()
+
+		innerRes := newEmptyPrometheusResponse()
+		inner := &mockHandler{}
+		inner.On("Do", mock.Anything, mock.Anything).Return(innerRes, nil)
```

Comment:
inner should never be called, right? So we can also remove this setup call. Otherwise a wrong behaviour of the cache can pass the test